### PR TITLE
1110 environment metrics fan speeds percent

### PR DESCRIPTION
**Pull request by lxwinspur**
1060: Redfish: Implement EnvironmentMetrics

- **Add FanSpeedsPercent**
  Upstream Link: https://gerrit.openbmc.org/c/openbmc/bmcweb/+/57715
  Changed-Id: I4cfc0aa28d68e7e0fa947251363deb6f06e36225
- **Add PowerWatts**
  Upstream Link: https://gerrit.openbmc.org/c/openbmc/bmcweb/+/57717
  Changed-Id: Ibe84a5e7fe0d2b232f925e457a094c021ca85d36
- **Add PowerLimitWantts:**
  Upstream Link: https://gerrit.openbmc.org/c/openbmc/bmcweb/+/47300
  Changed-Id: Iacd244e537ccab6572a0a55b7f30871774e097ff

upstream: https://gerrit.openbmc.org/q/topic:%22redfish-EnvironmentMetrics%22

### CODE REVIEW

**Comment by gtmills** [comment]
Jenkins run tests please

**Comment by gtmills** [comment]
@lxwinspur Can you add a description, what is the upstream commit?

**Comment by baemyung** [comment]
@lxwinspur 
As Gunnar suggested, please add Changed-Id to each of the commit message, and list gerrit urls here.

File: redfish-core/include/utils/fan_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+{
+    sdbusplus::message::object_path endpointPath{fanPath};
+    endpointPath /= "all_sensors";
+
+    dbus::utility::getAssociatedSubTree(
+        endpointPath,
+        sdbusplus::message::object_path("/xyz/openbmc_project/inventory"), 0,
+        sensorInterface,
+        [asyncResp, callback{std::forward<Callback>(callback)}](
+            const boost::system::error_code& ec,
```
Comment (gtmills) [inline redfish-core/include/utils/fan_utils.hpp:32]: Include error_code

File: redfish-core/include/utils/fan_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            for (const auto& [service, interfaces] : serviceMaps)
+            {
+                callback(service, sensorPath);
+            }
+        }
+    });
+}
+
+template <typename Callback>
+inline void getFanPaths(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
```
Comment (gtmills) [inline redfish-core/include/utils/fan_utils.hpp:57]: include memory

File: redfish-core/include/utils/fan_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                        const std::optional<std::string>& validChassisPath,
```
Comment (gtmills) [inline redfish-core/include/utils/fan_utils.hpp:58]: include optional

File: redfish-core/include/utils/fan_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                        Callback&& callback)
+{
+    sdbusplus::message::object_path endpointPath{*validChassisPath};
+    endpointPath /= "cooled_by";
+
+    dbus::utility::getAssociatedSubTreePaths(
+        endpointPath,
+        sdbusplus::message::object_path("/xyz/openbmc_project/inventory"), 0,
+        fanInterface,
+        [asyncResp, callback{std::forward<Callback>(callback)}](
```
Comment (gtmills) [inline redfish-core/include/utils/fan_utils.hpp:68]: ```#include <utility>```

File: redfish-core/lib/environment_metrics.hpp [hunk 5]
```diff
@@ -1,18 +1,101 @@
...
+{
+    redfish::fan_utils::getFanPaths(
+        asyncResp, validChassisPath,
+        [asyncResp, chassisId](
+            const dbus::utility::MapperGetSubTreePathsResponse& fanPaths) {
+        for (const auto& fanPath : fanPaths)
+        {
+            redfish::fan_utils::getFanSensorsObject(
+                asyncResp, fanPath,
+                std::bind_front(getFanSensorValue, asyncResp, chassisId));
```
Comment (gtmills) [inline redfish-core/lib/environment_metrics.hpp:95]: include functional

**Comment by lxwinspur** [comment]
> @lxwinspur As Gunnar suggested, please add Changed-Id to each of the commit message, and list gerrit urls here.

Updated

**Comment by lxwinspur** [comment]
> @lxwinspur Can you add a description, what is the upstream commit?

Updated

File: redfish-core/include/utils/fan_utils.hpp (hunk 1 above)
Comment (lxwinspur) [reply redfish-core/include/utils/fan_utils.hpp:32]: Done

File: redfish-core/include/utils/fan_utils.hpp (hunk 2 above)
Comment (lxwinspur) [reply redfish-core/include/utils/fan_utils.hpp:57]: Done

File: redfish-core/include/utils/fan_utils.hpp (hunk 3 above)
Comment (lxwinspur) [reply redfish-core/include/utils/fan_utils.hpp:58]: Done

File: redfish-core/lib/environment_metrics.hpp (hunk 5 above)
Comment (lxwinspur) [reply redfish-core/lib/environment_metrics.hpp:95]: Done

File: redfish-core/include/utils/fan_utils.hpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+#include "async_resp.hpp"
+#include "dbus_utility.hpp"
+#include "error_messages.hpp"
+
+#include <boost/system/error_code.hpp>
+
+#include <array>
+#include <memory>
+#include <optional>
+#include <string_view>
```
Comment (baemyung) [inline redfish-core/include/utils/fan_utils.hpp:12]:
```
#include <string>
```

File: redfish-core/lib/environment_metrics.hpp [hunk 7]
```diff
@@ -1,18 +1,102 @@
...
+            messages::internalError(asyncResp->res);
+            return;
+        }
+
+        updateFanSensorList(asyncResp, chassisId, fanSensorPath, value);
+    });
+}
+
+inline void
+    getFanSpeedsPercent(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
```
Comment (baemyung) [inline redfish-core/lib/environment_metrics.hpp:84]: It looks like - the upstream seems suggesting to restructure the code - "One callback per async action".
I also made a comment to it.

https://gerrit.openbmc.org/c/openbmc/bmcweb/+/57715/comment/8a18bdf9_f737ce8d/

**Comment by baemyung** [comment]
There seem several concerns on upstream commits.

@gtmills 
I'm not sure whether we should wait for the resolution of those upstream comments, or should we proceed to merge into 1110 (as they were in 1060) now and deal with any upstream updates later?

File: redfish-core/include/utils/fan_utils.hpp (hunk 6 above)
Comment (lxwinspur) [reply redfish-core/include/utils/fan_utils.hpp:12]: Done

File: redfish-core/include/utils/fan_utils.hpp (hunk 4 above)
Comment (lxwinspur) [reply redfish-core/include/utils/fan_utils.hpp:73]: Done

File: redfish-core/include/utils/fan_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                                Callback&& callback)
+{
+    sdbusplus::message::object_path endpointPath{fanPath};
+    endpointPath /= "all_sensors";
+
+    dbus::utility::getAssociatedSubTree(
+        endpointPath,
+        sdbusplus::message::object_path("/xyz/openbmc_project/inventory"), 0,
+        sensorInterface,
+        [asyncResp, callback{std::forward<Callback>(callback)}](
```
Comment (baemyung) [inline redfish-core/include/utils/fan_utils.hpp:36]: https://github.com/openbmc/bmcweb/commit/8cb2c024c4625e2fe2f0b107a865faffcd4bb770 suggests to use
```
[asyncResp, callback = std::forward<Callback>(callback)]
```

File: redfish-core/lib/environment_metrics.hpp [hunk 9]
```diff
@@ -1,18 +1,97 @@
...
+    nlohmann::json::object_t item;
+    item["SensorFanArrayExcerpt"]["DataSourceUri"] = boost::urls::format(
+        "/redfish/v1/Chassis/{}/Sensors/{}", chassisId, fanSensorName);
+    item["SensorFanArrayExcerpt"]["DeviceName"] = "Chassis #" + fanSensorName;
+    item["SensorFanArrayExcerpt"]["SpeedRPM"] = value;
+
+    nlohmann::json& fanSensorList =
+        asyncResp->res.jsonValue["FanSpeedsPercent"];
+    fanSensorList.emplace_back(std::move(item));
+    std::sort(fanSensorList.begin(), fanSensorList.end(),
```
Comment (baemyung) [inline redfish-core/lib/environment_metrics.hpp:45]: Upstreams would prefer to use `ranges`

```
    std::ranges::sort(fanSensorList,
              [](const nlohmann::json& c1, const nlohmann::json& c2) {
````

File: redfish-core/lib/environment_metrics.hpp [hunk 10]
```diff
@@ -1,18 +1,97 @@
...
+               c2["SensorFanArrayExcerpt"]["DataSourceUri"];
+    });
+    asyncResp->res.jsonValue["FanSpeedsPercent@odata.count"] =
+        fanSensorList.size();
+}
+
+inline void
+    getFanSensorValues(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                       const std::string& chassisId,
+                       const std::vector<std::pair<std::string, std::string>>&
```
Comment (baemyung) [inline redfish-core/lib/environment_metrics.hpp:57]: See my above comment to use `subtree`.

File: redfish-core/lib/environment_metrics.hpp [hunk 11]
```diff
@@ -41,6 +120,212 @@ inline void handleEnvironmentMetricsHead(
...
+        for (const auto& tempPath : paths)
+        {
+            sdbusplus::message::object_path path(tempPath);
+            std::string leaf = path.filename();
+            if (leaf == "total_power")
+            {
+                find = true;
+                break;
+            }
+        }
```
Comment (baemyung) [inline redfish-core/lib/environment_metrics.hpp:153]: Isn't his L143-L153 is a search like
```
auto find = std::ranges::find(paths, [][](const std::string &path){
            sdbusplus::message::object_path path(tempPath);
            std::string leaf = path.filename();
            return (leaf == "total_power");
     });

if (find = paths.end()) {
  ...
}
```

File: redfish-core/lib/fan.hpp [hunk 12]
```diff
@@ -207,7 +178,7 @@ inline void getValidFanPath(
     const std::function<void(const std::string& fanPath,
                              const std::string& service)>& callback)
 {
```
Comment (baemyung) [inline redfish-core/lib/fan.hpp:180]: Include
```
#include <functional>
```

File: redfish-core/lib/environment_metrics.hpp (hunk 10 above)
Comment (baemyung) [inline redfish-core/lib/environment_metrics.hpp:57]: Include
```
#include <vector>
```

File: redfish-core/lib/environment_metrics.hpp [hunk 13]
```diff
@@ -41,6 +120,212 @@ inline void handleEnvironmentMetricsHead(
...
+                messages::internalError(asyncResp->res);
+            }
+            return;
+        }
+
+        asyncResp->res.jsonValue["PowerLimitWatts"]["SetPoint"] = 0;
+        asyncResp->res.jsonValue["PowerLimitWatts"]["ControlMode"] =
+            "Automatic";
+
+        const uint32_t* powerCap = nullptr;
```
Comment (baemyung) [inline redfish-core/lib/environment_metrics.hpp:288]: Include
```
#include <cstdint>
```

**Comment by baemyung** [comment]
This will be redone by https://github.com/ibm-openbmc/bmcweb/pull/956

**Comment by gtmills** [comment]
Let's go with https://github.com/ibm-openbmc/bmcweb/pull/956. 
Closing this PR

**Pull request closed without merging**